### DISCUSSION TITLE
ESP32 Rebooting after a couple of days

### DISCUSSION THREAD

**Discussion by oleksiikutuzov** [HomeSpan/HomeSpan#337] [General]
Hello everyone,

this is rather a question to someone who better understands the ESP32, whether it can be a software or a hardware issue. I have a custom PCB for the ESP32-WROOM module to receive data from the CO2 sensor and IKEA air quality sensor. It sends it then to HomeKit via HomeSpan and has a web server for OTA updates and Prometheus metrics. It seems to be nothing special and not too much of a load. However, after some time it starts rebooting every couple of days. I made another unit for a friend of mine, it was working great for the first month or even more, and then suddenly started rebooting every day or so. The heap graph doesn't seem to show anything extraordinary. Maybe someone had this kind of an issue or has any ideas?
The source code is in the repository: https://github.com/oleksiikutuzov/esp32-homekit-air-quality
Also attaching the graphs

![Screenshot 2022-07-06 at 23 35 45](https://user-images.githubusercontent.com/8535871/177650209-71d67e32-0440-4651-a294-008210330ed9.png)
![Screenshot 2022-07-06 at 23 38 15](https://user-images.githubusercontent.com/8535871/177650205-190b3075-5c1b-4114-9cd3-90a96aaed9cf.png)

**Comment by HomeSpan** [comment]
Can I ask how you know it's rebooting?  I've recently found a bug in the weblog logic initialization routine - it is called whenever WiFi is established, which is correct upon initial startup but incorrect if WiFi was lost and then reconnected.  This causes two issues.  The first is that the boot time shown in the weblog output is incorrect. It's actually the time that WiFi was re-established.  So you can't rely on that to determine reboots. The second issue is that that if you start weblog with a time server a socket connection is reserved.  That means every time WiFi is lost a reconnected, another socket connection is reserved.  I need to fix this so the weblog init routine is called only once upon the first WiFi connection, and never on any subsequent reconnections.  

This second issue is also present in the OTA init routine.  It reserves a socket and is called every time WiFi is connected and re-connected.  

If the device has a stable WiFi connection, none of this matters.  But if the WiFi occasionally drops out, every time it reconnects another socket connection is reserved and thus can't be used for HomeKit requests.  If this gets too low I'm guessing that can cause a failure, which may lead to loss of HomeKit connectivity of a reboot.   

The fix is easy and I'll release a patch.  But to make sure this is the issue I'm testing a bunch of different scenarios as well as adding additional diagnostics, such  as the reason for a reboot. 

I have noticed that my ESP devices all lose WiFi connectivity at approximately the same time.  Often they are stable for days, and than the all simultaneously lose connectivity and need to re-establish.  They are all running different sketches, and sometimes only 1 or 2 devices drop and the other ones do not.  This indicates that the the drops are not due to the ESP software or hardware, but are real drops that I need to handle in the code more gracefully.

**Reply by oleksiikutuzov** [reply]
I am not using the weblog or built-in OTA, so this should not be this particular issue. I am using ESPAsyncWebServer, AsyncTCP, and AsyncElegantOTA for the server. It might of course also be that they have the same bug. 

For uptime count, I am using the `esp_timer_get_time()` function of the ESP, so this should provide the correct time since the last reboot. And also the CO2 sensor has a warm-up period of 3 minutes on startup, during which it is blinking with LED, so it is noticeable.

The WiFi connection also should not be a problem – the device is standing in the same room about 3 meters away from the router, but who knows. Why I mentioned the hardware, I was also trying to use the generic ESP32 modules from Aliexpress in this particular device. I always could not connect to the OTA server after some time and I guess it got unavailable in HomeKit, which is not the case for the genuine modules from Espressif. They are working fine (well, at least the server is always available). I don't know what is the difference between them and whether this could be an unstable WiFi connection and reconnects causing reserving extra sockets, but it was the case. However, those modules are working fine with boards controlling the Neopixel strips, this is weird. 

I can also see, that devices are dropping from the HomeKit, but was usually blaming the router. I have two boards controlling the LED strips and this sensor. Lately, the LED controllers are unavailable for quite a long time and then at some point, they might get back to HomeKit, but I am not monitoring their uptime at the moment to say something. But this sensor is always online, it is updating the Prometheus metrics every 30 seconds and I can see that it is stable even if it gets unavailable in HomeKit.

**Reply by HomeSpan** [reply]
May want to try logging the reboot reason code - might provide some insight into what happens before the reboot.

**Reply by oleksiikutuzov** [reply]
Yeah, I somehow didn't even think of this. Will do and report back if there's something interesting, thanks

**Reply by oleksiikutuzov** [reply]
The reason it provides is `SW_CPU_RESET`, which isn't quite informative in this case. But during those 48 hours, there were some spikes and after reboot, it freed about 2 kB.

<img width="512" alt="Screenshot 2022-07-09 at 15 09 20" src="https://user-images.githubusercontent.com/8535871/178107177-143e94b7-2df1-4b5f-946f-c484c2ee63d2.png">

**Comment by HomeSpan** [comment]
Which reason code functionality are you using. I would try [esp_reset_reason()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/misc_system_api.html#_CPPv418esp_reset_reason_t).  It can distinguish between a hardware reset (e.g. power drop or reset button), a software reset (e.g. ESP.restart()), a kernel panic, and a watchdog timer.

**Reply by oleksiikutuzov** [reply]
I was using `rtc_get_reset_reason()`. Will try this one now, thanks

**Reply by oleksiikutuzov** [reply]
I caught reboot reason 6, which should be SDIO_RESET (Reset by SLC module, reset digital core). Couldn't find any meaningful explanation of what this is and what can cause this. Seems more like a hardware issue, so I just hooked a dev module on a breakout board instead of a diy module and will watch how it behaves now.

upd. Just checked the dev module, it also has the same reboot reason 6, just after 1.5 days of running.

**Reply by HomeSpan** [reply]
I think a reset reason code of 6 is a task watchdog timeout.

**Reply by oleksiikutuzov** [reply]
Yes, right, I checked the wrong reset reasons again. This changes the thing, seems like this might be an issue with AsyncWebServer. I will report again after some time

**Comment by oleksiikutuzov** [comment]
I can now say that the reason was the ESPAsyncWebServer library triggering the task watchdog. Running for almost 2 weeks with no issues using the default WebServer. Thanks for your suggestions and comments!

**Reply by HomeSpan** [reply]
Really glad to hear it seems to be working.  Thanks for the update.
